packages/runtime/src/interpreter/opcode-list.ts

```diff
@@ -5087,3 +5091,112 @@ export class Replace3 extends Replace {
 		return super.execute(stack);
 	}
 }
+
+/**
+ * Opcode: json_ref r
+ * Stack: ..., A: []byte, B: []byte → ..., any
+ * return key B's value from a valid utf-8 encoded json object A
+ */
+export class Json_ref extends Op {
+	readonly line: number;
+	readonly jsonType: string;
+	length = 1;
+
+	/**
+	 * Asserts 1 argument is passed.
+	 * @param args Expected arguments: [e], where e = {JSONString, JSONUint64 and JSONObject}.
+	 * @param line line number in TEAL file
+	 */
+	constructor(args: string[], line: number) {
+		super();
+		this.line = line;
+		assertLen(args.length, 1, line);
+		const argument = args[0];
+		switch (argument) {
+			case json_refTypes.JSONString: {
+				this.jsonType = "byte";
+				break;
+			}
+			case json_refTypes.JSONUint64: {
+				this.jsonType = "int";
+				break;
+			}
+			case json_refTypes.JSONObject: {
+				this.jsonType = "object";
+				break;
+			}
+			default: {
+				throw new RuntimeError(RUNTIME_ERRORS.TEAL.UNKNOWN_JSON_TYPE, {
+					jsonType: argument,
+					line: this.line,
+				});
+			}
+		}
+	}
+
+	computeCost(): number {
+		return 25 + 2 * Math.ceil(this.length / 7); // cost = 25 + ceil(bytes / 7)
+	}
+	
+	execute(stack: TEALStack): number {
+		this.assertMinStackLen(stack, 2, this.line);
+		const key = this.assertBytes(stack.pop(), this.line);
+		const object = this.assertBytes(stack.pop(), this.line);
+		this.length = object.length;
+		const enc = new TextDecoder("utf-8");
+		const decoded = enc.decode(object);
```

Comment:
lets follow the same convention both for object and the key. Maybe 'decodedObj' and 'decodedKey'?